doc/_build/man/posty.1
.\" Man page generated from reStructuredText.
.
.TH "POSTY" "1" "Jan 20, 2018" "2.0" "Posty"
.SH NAME
posty \- Posty Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.sp
A simple static site generator tool. It reads in a series of posts and pages
containing YAML metadata and Markdown text, and renders them as HTML.
.SH QUICKSTART
.sp
hi.
.SH CLI REFERENCE
.SH CONFIG FILE
.SH POST AND PAGE SCHEMA
.SH TEMPLATING
.SH IMPORTING FROM OTHER STATIC SITE GENERATORS
.SH DEVELOPMENT
.SH POSTY
.SS posty package
.SS Subpackages
.SS posty.renderer package
.SS Submodules
.SS posty.renderer.atom module
.INDENT 0.0
.TP
.B class posty.renderer.atom.AtomRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.feed.FeedRenderer\fP
.sp
Renderer that outputs an Atom feed XML file
.INDENT 7.0
.TP
.B filename = \(aqatom.xml\(aq
.UNINDENT
.INDENT 7.0
.TP
.B output()
Output the Atom feed file
.UNINDENT
.INDENT 7.0
.TP
.B url()
Return the URL to this feed file
.UNINDENT
.UNINDENT
.SS posty.renderer.base module
.INDENT 0.0
.TP
.B class posty.renderer.base.Renderer(site, output_path=\(aqbuild\(aq)
Bases: \fBobject\fP
.sp
Base class that all renderers inherit off of. Each child class must
implement \fBrender_site()\fP with their own rendering logic.
.INDENT 7.0
.TP
.B ensure_output_path()
Ensure that the output directory \fBself.output_path\fP exists
.UNINDENT
.INDENT 7.0
.TP
.B render_site()
.UNINDENT
.UNINDENT
.SS posty.renderer.feed module
.INDENT 0.0
.TP
.B class posty.renderer.feed.FeedRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.base.Renderer\fP
.sp
Base class for all feed Renderers (RSS, Atom)
.INDENT 7.0
.TP
.B output()
This method must be implemented by child classes. It gets called during
render_site to output the specific file, such as the RSS file or Atom
file
.UNINDENT
.INDENT 7.0
.TP
.B render_posts()
Add each post to the feed
.UNINDENT
.INDENT 7.0
.TP
.B render_site()
.UNINDENT
.INDENT 7.0
.TP
.B url()
Return the URL to this feed file
.UNINDENT
.UNINDENT
.SS posty.renderer.html module
.INDENT 0.0
.TP
.B class posty.renderer.html.HtmlRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.base.Renderer\fP
.sp
Renderer that outputs HTML files
.INDENT 7.0
.TP
.B prepare_content()
Do a first\-pass rendering of each post and page text, treating the text
as Jinja2 templates. This lets us use basic jinja in the markdown to
be able to use filter functions like this:
.sp
\fB{{ "img/cool_pic.jpg" | media_url }}\fP
.UNINDENT
.INDENT 7.0
.TP
.B render_page(page, template_name=\(aqpage.html\(aq)
.INDENT 7.0
.TP
.B Parameters
\fBpage\fP – a Page object
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B render_post(post, template_name=\(aqpost.html\(aq)
.INDENT 7.0
.TP
.B Parameters
\fBpost\fP – a Post object
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B render_posts(posts, prefix=\(aq\(aq, template_name=\(aqposts.html\(aq)
Render a list of posts as sets of pages where each page has
\fBnum_posts_per_page\fP posts. Each page of posts will be rendered to
the path page/:page/index.html relative to the Renderer output_path
.sp
If \fBprefix\fP is given, add that will be put in between the output_path
and page path. For example if the prefix is ‘tags/foo/’ then a page
path would look like ‘tags/foo/page/:page/index.html’
.UNINDENT
.INDENT 7.0
.TP
.B render_site()
Given a Site object, render all of its components
.INDENT 7.0
.TP
.B Parameters
\fBsite\fP – a loaded Site object
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B render_site_posts()
Renders all of the multi\-post pages, N per page
.UNINDENT
.INDENT 7.0
.TP
.B render_site_tags(template_name=\(aqposts.html\(aq)
Renders all of the per\-tag multi\-post pages, N per page
.UNINDENT
.UNINDENT
.SS posty.renderer.json module
.INDENT 0.0
.TP
.B class posty.renderer.json.JsonRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.base.Renderer\fP
.sp
Renderer that outputs a JSON representation of the Site to \fBsite.json\fP
within the output directory
.INDENT 7.0
.TP
.B render_site()
Render the Site to \fBsite.json\fP
.UNINDENT
.UNINDENT
.SS posty.renderer.rss module
.INDENT 0.0
.TP
.B class posty.renderer.rss.RssRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.feed.FeedRenderer\fP
.sp
Renderer that outputs a RSS feed XML file
.INDENT 7.0
.TP
.B filename = \(aqrss.xml\(aq
.UNINDENT
.INDENT 7.0
.TP
.B output()
Output the RSS feed file
.UNINDENT
.INDENT 7.0
.TP
.B url()
Return the URL to this feed file
.UNINDENT
.UNINDENT
.SS posty.renderer.util module
.INDENT 0.0
.TP
.B posty.renderer.util.absolute_url_func(site)
Returns a markdown filter function that returns an absolute URL for the
given relative URL, simply concatenating config[‘base_url’] with the URL.
.UNINDENT
.INDENT 0.0
.TP
.B posty.renderer.util.markdown(text)
Returns the rendered version of the given Markdown text
.UNINDENT
.INDENT 0.0
.TP
.B posty.renderer.util.media_url_func(site)
Returns a filter function that returns a full media URL for the given file,
scoped to the given Site object.
.sp
For example, if the Site has its base_url set to ‘/foo/’ then:
img/my_picture.jpg \-> /foo/media/img/my_picture.jpg
.UNINDENT
.SS Module contents
.INDENT 0.0
.TP
.B class posty.renderer.HtmlRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.base.Renderer\fP
.sp
Renderer that outputs HTML files
.INDENT 7.0
.TP
.B prepare_content()
Do a first\-pass rendering of each post and page text, treating the text
as Jinja2 templates. This lets us use basic jinja in the markdown to
be able to use filter functions like this:
.sp
\fB{{ "img/cool_pic.jpg" | media_url }}\fP
.UNINDENT
.INDENT 7.0
.TP
.B render_page(page, template_name=\(aqpage.html\(aq)
.INDENT 7.0
.TP
.B Parameters
\fBpage\fP – a Page object
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B render_post(post, template_name=\(aqpost.html\(aq)
.INDENT 7.0
.TP
.B Parameters
\fBpost\fP – a Post object
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B render_posts(posts, prefix=\(aq\(aq, template_name=\(aqposts.html\(aq)
Render a list of posts as sets of pages where each page has
\fBnum_posts_per_page\fP posts. Each page of posts will be rendered to
the path page/:page/index.html relative to the Renderer output_path
.sp
If \fBprefix\fP is given, add that will be put in between the output_path
and page path. For example if the prefix is ‘tags/foo/’ then a page
path would look like ‘tags/foo/page/:page/index.html’
.UNINDENT
.INDENT 7.0
.TP
.B render_site()
Given a Site object, render all of its components
.INDENT 7.0
.TP
.B Parameters
\fBsite\fP – a loaded Site object
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B render_site_posts()
Renders all of the multi\-post pages, N per page
.UNINDENT
.INDENT 7.0
.TP
.B render_site_tags(template_name=\(aqposts.html\(aq)
Renders all of the per\-tag multi\-post pages, N per page
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B class posty.renderer.JsonRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.base.Renderer\fP
.sp
Renderer that outputs a JSON representation of the Site to \fBsite.json\fP
within the output directory
.INDENT 7.0
.TP
.B render_site()
Render the Site to \fBsite.json\fP
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B class posty.renderer.RssRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.feed.FeedRenderer\fP
.sp
Renderer that outputs a RSS feed XML file
.INDENT 7.0
.TP
.B filename = \(aqrss.xml\(aq
.UNINDENT
.INDENT 7.0
.TP
.B output()
Output the RSS feed file
.UNINDENT
.INDENT 7.0
.TP
.B url()
Return the URL to this feed file
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B class posty.renderer.AtomRenderer(site, output_path=\(aqbuild\(aq)
Bases: \fBposty.renderer.feed.FeedRenderer\fP
.sp
Renderer that outputs an Atom feed XML file
.INDENT 7.0
.TP
.B filename = \(aqatom.xml\(aq
.UNINDENT
.INDENT 7.0
.TP
.B output()
Output the Atom feed file
.UNINDENT
.INDENT 7.0
.TP
.B url()
Return the URL to this feed file
.UNINDENT
.UNINDENT
.SS Submodules
.SS posty.cli module
.SS posty.config module
.INDENT 0.0
.TP
.B class posty.config.Config(path=\(aqconfig.yml\(aq)
Bases: \fBcollections.abc.MutableMapping\fP
.sp
Config object that gets passed around to various other objects. Loads
config from a given YAML file.
.INDENT 7.0
.TP
.B Parameters
\fBpath\fP – Path to a YAML file to read in as config
.UNINDENT
.INDENT 7.0
.TP
.B clean_config()
Validate and clean the already\-loaded config
.UNINDENT
.INDENT 7.0
.TP
.B load()
Load the YAML config from the given path, return the config object
.UNINDENT
.UNINDENT
.SS posty.exceptions module
.INDENT 0.0
.TP
.B exception posty.exceptions.InvalidConfig(config_obj, reason)
Bases: \fI\%posty.exceptions.PostyError\fP
.UNINDENT
.INDENT 0.0
.TP
.B exception posty.exceptions.InvalidObject
Bases: \fI\%posty.exceptions.PostyError\fP
.UNINDENT
.INDENT 0.0
.TP
.B exception posty.exceptions.MalformedInput
Bases: \fI\%posty.exceptions.PostyError\fP
.UNINDENT
.INDENT 0.0
.TP
.B exception posty.exceptions.PostyError
Bases: \fBRuntimeError\fP
.UNINDENT
.INDENT 0.0
.TP
.B exception posty.exceptions.UnableToImport
Bases: \fI\%posty.exceptions.PostyError\fP
.UNINDENT
.SS posty.importers module
.sp
Functions to import from various other static site generators
.INDENT 0.0
.TP
.B class posty.importers.Importer(site, src_path)
Bases: \fBposty.model.ABC\fP
.sp
Base class for all importers
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBsite\fP – Site object for the destination
.IP \(bu 2
\fBsrc_path\fP – Path to the thing to import
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B ensure_directories()
.UNINDENT
.INDENT 7.0
.TP
.B run()
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B class posty.importers.Posty1Importer(site, src_path)
Bases: \fI\%posty.importers.Importer\fP
.sp
Importer to pull from a Posty 1.x site
.INDENT 7.0
.TP
.B import_media()
.UNINDENT
.INDENT 7.0
.TP
.B import_pages()
.UNINDENT
.INDENT 7.0
.TP
.B import_posts()
.UNINDENT
.INDENT 7.0
.TP
.B import_templates()
.UNINDENT
.INDENT 7.0
.TP
.B run()
.UNINDENT
.UNINDENT
.SS posty.model module
.INDENT 0.0
.TP
.B class posty.model.ABC
Bases: \fBobject\fP
.UNINDENT
.INDENT 0.0
.TP
.B class posty.model.Model(payload, config=None)
Bases: \fI\%posty.model.ABC\fP, \fBcollections.abc.MutableMapping\fP
.sp
Base class for objects representing things stored as YAML, such as a Post
or a Page
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBpayload\fP – A dict representing the backing payload for this object
.IP \(bu 2
\fBconfig\fP – A Config object
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B as_dict()
Return a true dict representation of this object, suitable for
serialization into JSON or YAML
.UNINDENT
.INDENT 7.0
.TP
.B classmethod from_yaml(file_contents, config=None)
Load an object from its YAML file representation
.UNINDENT
.INDENT 7.0
.TP
.B path_on_disk()
Returns the relative path on disk to the object, for rendering purposes
.UNINDENT
.INDENT 7.0
.TP
.B url()
Returns the URL path to this resource
.UNINDENT
.INDENT 7.0
.TP
.B validate()
This should be implemented by the child class to verify that all fields
that are expected exist on the payload, and set any that aren’t
.UNINDENT
.UNINDENT
.SS posty.page module
.INDENT 0.0
.TP
.B class posty.page.Page(payload, config=None)
Bases: \fBposty.model.Model\fP
.sp
Representation of a page
.INDENT 7.0
.TP
.B classmethod from_yaml(file_contents, config=None)
Return a Page from the given file_contents
.UNINDENT
.INDENT 7.0
.TP
.B path_on_disk()
.UNINDENT
.INDENT 7.0
.TP
.B url()
.UNINDENT
.INDENT 7.0
.TP
.B validate()
.UNINDENT
.UNINDENT
.SS posty.post module
.INDENT 0.0
.TP
.B class posty.post.Post(payload, config=None)
Bases: \fBposty.model.Model\fP
.sp
Representation of a post
.INDENT 7.0
.TP
.B classmethod from_yaml(file_contents, config=None)
Returns a Post from the given file_contents
.UNINDENT
.INDENT 7.0
.TP
.B path_on_disk()
.UNINDENT
.INDENT 7.0
.TP
.B url()
.UNINDENT
.INDENT 7.0
.TP
.B validate()
.UNINDENT
.UNINDENT
.SS posty.site module
.INDENT 0.0
.TP
.B class posty.site.Site(site_path=\(aq.\(aq, config_path=None)
Bases: \fBobject\fP
.sp
Representation of an entire site with posts and pages. This is the main
class that conrols everything.
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBsite_path\fP – Path to the directory containing site content (pages, posts, templates)
.IP \(bu 2
\fBconfig_path\fP – Path to the config file, defaults to \fB$SITE_PATH/config.yml\fP
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B config
Returns this site’s config as read from the config file
.UNINDENT
.INDENT 7.0
.TP
.B copyright
Returns a string of the copyright info, based on the configured author
and the years of the first and last post
.UNINDENT
.INDENT 7.0
.TP
.B init()
Initialize a new Posty site at the given path
.UNINDENT
.INDENT 7.0
.TP
.B load()
Load the site from files on disk into our internal representation
.UNINDENT
.INDENT 7.0
.TP
.B page(slug)
Returns a Page object by its slug
.INDENT 7.0
.TP
.B Parameters
\fBslug\fP – slug of the page to find
.TP
.B Returns
A page dict
.TP
.B Raises
\fBPostyError\fP – if no page could be found
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B post(slug)
Returns a Post object by its slug
.INDENT 7.0
.TP
.B Parameters
\fBslug\fP – slug of the post to find
.TP
.B Returns
A post dict
.TP
.B Raises
\fBPostyError\fP – if no post could be found
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B render(output_path=\(aqbuild\(aq)
Render the site with the various renderers
.INDENT 7.0
.IP \(bu 2
HTML
.IP \(bu 2
JSON
.IP \(bu 2
RSS (if \fBfeeds.rss\fP is True in the config)
.IP \(bu 2
Atom (if \fBfeeds.atom\fP is True in the config)
.UNINDENT
.UNINDENT
.UNINDENT
.SS posty.util module
.sp
Various utility functions
.INDENT 0.0
.TP
.B posty.util.bucket(_list, size)
Bucket the list \fB_list\fP into chunks of up to size \fBsize\fP
.sp
Example:
bucket([1,2,3,4,5], 2) \-> [[1,2], [3,4], [5]]
.UNINDENT
.INDENT 0.0
.TP
.B posty.util.slugify(text)
Returns a slugified version of the given \fBtext\fP
.UNINDENT
.INDENT 0.0
.TP
.B posty.util.slugify_posty1(text)
Returns a Posty 1.x compatible slugified version of \fBtext\fP
.UNINDENT
.SS Module contents
.INDENT 0.0
.IP \(bu 2
genindex
.IP \(bu 2
modindex
.IP \(bu 2
search
.UNINDENT
.SH AUTHOR
Nick Pegg
.SH COPYRIGHT
2018, Nick Pegg
.\" Generated by docutils manpage writer.
.


posty.py
#!/usr/bin/python

'''
Posty
Written by Nick Pegg
http://github.com/nickpegg/posty/

Very simple static site generator.
Takes in markdown files, spits out a website.
'''

# Various settings
SITE_TITLE = "My Website"
SITE_URL = "http://example.com/"
SITE_DESC = "My website of stuff and things"

PER_PAGE = 10        # How many posts should there be per page?

# Locations of various files that Posty needs
# All paths must end with a slash
TEMPLATE_PATH = '_templates/'
POSTS_PATH = '_posts/'
PAGES_PATH = '_pages/'
MEDIA_PATH = '_media/'

OUTPUT_PATH = 'output/'


# Only edit stuff below here if you're feeling adventurous :)

import os
import sys
import shutil
import datetime
import re
from operator import itemgetter

import sqlite3
import markdown2
import yaml
import PyRSS2Gen as rss2
from jinja2 import Environment, FileSystemLoader


# Create a jinja environment
jenv = Environment(loader=FileSystemLoader(TEMPLATE_PATH), cache_size=0)


## Utility functions ###
def render_markdown(value):
    ''' Simple jinja filter to render markdown text '''
    return markdown2.markdown(unicode(value))
jenv.filters['markdown'] = render_markdown

def file_write(filename, html):
    try:
        file = open(filename, 'w')
        file.write(html)
    except Exception as e:
        print("Unable to write file: " + str(e))
        
def read_header(file):
    item = None
    
    contents = file.read()
    
    try:
        matches = re.match('(---(?:\r?\n)+(?:.*(?:\r?\n))+)---(?:\r?\n)((.*(?:\r?\n)?)+)', contents, re.MULTILINE).groups()
        body = ''
        
        header = matches[0]
        for m in matches[1:]:
            body += m
            
        item = yaml.load(header)
    except Exception as e:
        print("The header in " + file.name + " is screwey. Fix it.")
        print("Error: " + str(e))
        quit(-1)
        
    item['body'] = body
    
    return item
    
def slugify(text):
    """ This hardly covers many cases, but it covers what I've come across.
    HELLLOOOOO lazy programmer. """
    return str(text).strip().lower().replace(' ', '_').replace('#', '_')
    

### Main functions ###
def render():
    """ Render the posts and pages into a bunch of HTML files """
    
    if os.path.isdir(OUTPUT_PATH):
        try:
            shutil.rmtree(OUTPUT_PATH)
        except Exception as e:
            print("Unable to delete and recreate output folder. Please delete it by hand and try again.")
            print("Error: " + str(e))
            quit(-1)
    os.mkdir(OUTPUT_PATH)
    
    # Gather up the posts
    posts = list()
    
    if os.path.isdir('_posts'):
        for file in os.listdir('_posts'):
            post = read_header(open('_posts/' + file))
            
            # Might as well validate while we're here
            if post['body'] == '':
                print("Error: " + file + " has no post body. Skipping.")
                continue
            if post.get('title') is None:
                print("Error: " + file + " has no title. Skipping.")
                continue
            if post.get('date') is None:
                print("Error: " + file + " has no date. Skipping.")    
                continue
                
            # Everything must have gone fine, so let's add the post.    
            posts.append(post)
    
    # Gather up the pages
    pages = list()
    
    if os.path.isdir('_pages'):
        for file in os.listdir('_pages'):
            page = read_header(open('_pages/' + file))
            
            # Let's validate this guy
            if page['body'] == '':
                print("Error: " + file + " has no page body. Skipping.")
                continue
            if page.get('title') is None:
                print("Warning: " + file + " has no title. You may want one depending on your template.")
            if page.get('url') is None:
                print("Warning: " + file + " doesn't specify a URL. Using the page filename...")
                page['url'] = file.replace('.yaml', '.html').replace('.yml', '.html')
                
            # Everything went better than expected.
            pages.append(page)
    
    # Let's sort these pages out so they show up right
    # First in alpha order, then by parent/child
    # Children with a non-existant parent get orphaned. Poor bastards.
    sorted_pages = sorted(pages, key=itemgetter('title'))

    pages = list()
    for page in sorted_pages:
        if page.get('parent') is None:
            pages.append(page)

            # Stupid programming, but screw it, I'm only going one child deep.
            for possible_child in sorted_pages: 
                parent = possible_child.get('parent')
                
                if parent is not None and parent == page.get('title'):
                    if possible_child in pages:
                        pages.remove(possible_child)
                        
                    pages.append(possible_child)

    # Let's sort the posts by date
    posts = sorted(posts, key=itemgetter('date'))
    posts.reverse()
    
    # Render the posts as individual pages
    template = jenv.get_template('post.html')
    for post in posts:
        date = post['date']
        post['url'] = str(date.year) + "/" + str(date.month) + "/" + slugify(post['title']) + ".html"
        
        # Let's make sure that the directories exist
        if not os.path.isdir(OUTPUT_PATH + os.path.dirname(post['url'])):
            os.makedirs(OUTPUT_PATH + os.path.dirname(post['url'])) 
        try:
            html = template.render(post=post, pages=pages)
        except Exception as e:
            print("Problem rendering the post " + str(post.get('title')))
            print(str(e))

        file_write(OUTPUT_PATH + post['url'], html)

    # Render the posts as grouped pages
    template = jenv.get_template('posts.html')
    for start in range(0, len(posts), PER_PAGE):
        prev_page = "posts" + str(start - PER_PAGE) + ".html"
        next_page = "posts" + str(start + PER_PAGE) + ".html"

        if start == 0:
            prev_page = None

        if len(posts) - start <= PER_PAGE:
            next_page = None

        
        html = template.render(posts=posts[start:(start+PER_PAGE-1)],
            pages=pages, prev_page=prev_page, next_page= next_page)
        
        
        if start == 0:
            file_write(OUTPUT_PATH + "index.html", html)
            file_write(OUTPUT_PATH + "posts" + str(start) + ".html", html)
        else:
            file_write(OUTPUT_PATH + "posts" + str(start) + ".html", html)

    # Render an RSS feed of the last 10 posts
    rss_items = list()
    for post in posts[:10]:
        item = rss2.RSSItem(
	    title = post.get('title'),
            link = SITE_URL + post.get('url'),
            guid = rss2.Guid(SITE_URL + post.get('url')),
            pubDate = datetime.datetime.combine(post.get('date'), datetime.time())
	)
        rss_items.append(item)
		
    feed = rss2.RSS2(
        title = SITE_TITLE,
        link = SITE_URL,
        description = SITE_DESC,
        lastBuildDate = datetime.datetime.now(),
        items = rss_items
    )
	
    feed.write_xml(open(OUTPUT_PATH + 'feed.rss', 'w'))

    # Render the pages
    template = jenv.get_template('page.html')
    for page in pages:
        if page['url'][-1:] == '/':
            page['url'] += 'index.html'
            
        # Let's make sure that the directories exist
        if not os.path.isdir(OUTPUT_PATH + os.path.dirname(page['url'])):
            os.makedirs(OUTPUT_PATH + os.path.dirname(page['url']))

        try:
            html = template.render(page=page, pages=pages)
        except Exception as e:
            print("Something bad happened while trying to render page " + str(page.get('title')))
            print(str(e))

        file_write(OUTPUT_PATH + page['url'], html)
        
    # Finally, copy all static content over
    shutil.copytree('_media', 'output/media/')

    print("\nDone rendering! Check out output/index.html\n")

if __name__ == "__main__":
    render()
    quit(0)
